### PR TITLE
TypeScript

### PR DESCRIPTION
**Pull request by cniska**
After facing some hard-to-solve issues with Flowtype in a larger project one of our teams were working on, we decided that it would make sense to replace Flowtype with TypeScript.

We know that TypeScript's type system is not as robust Flowtype's but we felt that we were spending way too much of the development time with typing our code. There is close to no gain from a tool that makes development more frustrating when its promise is the opposite. Also Flowtype doesn't have a public roadmap, which makes it really hard to plan ahead.

I took the following steps to make this happen:
- Remove everything related to Flow
- Create a new project using Microsoft's TypeScript flavor of Create React App
- Copy over all necessary files and configurations
- Convert all `.js` files to `.tsx` files
- Change to use TSLint instead of ESLint
- Fix build, tests and CI
- Some minor tweaks

There are still a few things that need to be decided before merging:
- ~~Should we use the `ts-loader` or `awesome-typescript-loader`?~~
- ~~Is way CSS module are handled adequate or should it still be improved?~~
- ~~Should the definition files be committed or ignored?~~
- Is HMR working as it's supposed to?
- Should we include [mobx-state-tree](https://github.com/mobxjs/mobx-state-tree)?

### CODE REVIEW

File: config/webpack.config.dev.js [hunk 1]
```diff
@@ -142,18 +149,11 @@ module.exports = {
...
             include: paths.appSrc,
-            loader: require.resolve('babel-loader'),
-            options: {
-              
-              // This is a feature of `babel-loader` for webpack (not Babel itself).
-              // It enables caching results in ./node_modules/.cache/babel-loader/
-              // directory for faster rebuilds.
-              cacheDirectory: true,
-            },
+            loader: require.resolve('ts-loader'),
```
Comment (villeristi) [inline config/webpack.config.dev.js]: I'd suggest using https://github.com/s-panferov/awesome-typescript-loader instead

Comment (cniska) [reply config/webpack.config.dev.js]: Ok, this is based on https://github.com/Microsoft/TypeScript-React-Starter#typescript-react-starter

Comment (villeristi) [reply config/webpack.config.dev.js]: In my opinion the atl goes better with babel. But, I've not played with it more than one project so the opinion is a _bit_ narrow.

Comment (cniska) [reply config/webpack.config.dev.js]: Let's keep it like this for now because it does work.

**Comment by cniska** [comment]
@villeristi Can you check that the types are defined correctly?

File: config/webpack.config.dev.js (hunk 1 above)
Comment (AriaMinaei) [reply config/webpack.config.dev.js]: I'd add `options: {transpileOnly: true}` so type errors don't get in the way.

Comment (cniska) [reply config/webpack.config.dev.js]: @AriaMinaei where should I add that? We have two different loaders configured in the webpack for TypeScript.

**Comment by cniska** [comment]
@villeristi @AriaMinaei should generated `.d.ts` files be commited? For example those generated for `.css`-files.

**Comment by cniska** [comment]
I'm now happy with the state of this branch. 

@villeristi @AriaMinaei could you do a complete review on the changes and point out if I missed something? 

Thanks in advance.

File: config/webpack.config.dev.js (hunk 1 above)
Comment (AriaMinaei) [reply config/webpack.config.dev.js]: This is my loaders for `/\.tsx?$/` files for development: 
```
         {
            test: /\.(tsx?)$/,
            include: paths.appSrc,
            use: [
              {
                loader: require.resolve('babel-loader'),
                options: {
                  babelrc: true,
                  plugins: ['react-hot-loader/babel'],
                },
              },
              {
                loader: require.resolve('ts-loader'),
                options: {
                  // useTranspileModule: true,
                  transpileOnly: true,
                },
              },
            ],
          },
```

File: src/hello/components/Hello.tsx [hunk 2]
```diff
@@ -1,12 +1,10 @@
...
-import {observer} from 'mobx-react'
-import css from './Hello.css'
+import * as React from 'react'
+import { observer } from 'mobx-react'
 import helloStore from '../store'
+import * as css from './Hello.css'
 
 @observer
-export class Hello extends Component {
+export class Hello extends React.Component {
```
Comment (AriaMinaei) [inline src/hello/components/Hello.tsx]: I think a good convention would be for each react component to specify an `interface Props {}` and `interface State {}`, and then `export class C extends React.Component<Props, State>`

**Comment by AriaMinaei** [comment]
Other than my last comments, it's all LGTM.

Off-topic: Consider switching to mobx-state-tree rather than bare mobx. I've found it makes everything easier (snapshots, saving to localStorage, undo/redo, being able to use redux devtools, etc)

**Comment by AriaMinaei** [comment]
Committing the `.css.d.ts` files doesn't cause trouble in practice, and also allows a tyepecript typecheck pass to be doable in the CI without having to run webpack first.

**Comment by cniska** [comment]
I tried to both ignore and get rid of the `.css.d.t` definition files but neither really worked. I'm still not satisfied with the state of this branch but I'm also facing a wall in this matter.

**Comment by cniska** [comment]
@AriaMinaei bump

**Comment by AriaMinaei** [comment]
Added a commit, but the build is still failing. I'll take a second look later this week.

**Comment by cniska** [comment]
Ok thanks @AriaMinaei.

**Comment by villeristi** [comment]
Addded couple of commits, basically refactor things, introduce tslint-type-checker on separate process & clean up.

**Comment by suppayami** [comment]
Should we remove the `flow` check in `ci` script?

File: package.json [hunk 3]
```diff
@@ -60,7 +60,7 @@
     "webpack-dev-server": "2.7.1"
   },
   "scripts": {
-    "ci": "CI=true run-p test",
+    "ci": "CI=true run-p flow lint-staged test:cov",
```
Comment (suppayami) [inline package.json]: flow is not used here

**Comment by suppayami** [comment]
Another hack for importing `.css` is making a definition file declaring *.css module, though this won't help with cssModules. This also supports importing other media files (.png etc...)

```typescript
declare module '*.css'
declare module '*.png'
```

**Comment by hugovk** [comment]
Closing, this project is deprecated. Please use [Create React App](https://github.com/facebook/create-react-app) instead.

**Pull request closed without merging**